pagefaultgames/pokerogue
# [Bug] Introducing a check for pokémon being on field to activate abilities

## What are the changes the user will see?
Abilities such as Neutralizing Gas stop working after the user is out of the field, whether this is due to switching, a move such as roar, or an ability such as wimp out.
Additionally, Future Sight will now only hit once when used by a Pokémon with Parental Bond, if the user switches out before Future Sight hits (intended behavior).


## Why am I making these changes?
It is well known that the effect of Neutralizing Gas persists after a mon is hit by a forced switch move, see e.g. discord issue
https://discord.com/channels/1125469663833370665/1320888834334130186
The change brings the gameplay in line with player expectations.


## What are the changes from a developer perspective?
In the function canApplyMove(...), a new check is introduced to confirm that the ability user is on the field.

A few tests failed, because of not correctly adding mons to game.scene.field (Effectiveness) or because of checking if a mon not on the field has an ability (Pastel Veil). As far as my understanding goes, the issue in this case is within the tests themselves, and they have been changed accordingly. In particular, hasAbility is only used on mons on the field elsewhere, which suggests that is the only intended use case.

The test for Parental Bond interacting with Future Sight when the user leaves the field has been updated to reflect the expected result (according to a TODO in the test itself).

Tests for abilities that increase the rates of double battles have been refactored to be more robust and deterministic. The previous versions were liable to return a positive due to naturally generating a double battle rather than due to the ability itself.

The only situation where tests actually failed is for abilities which activate before switching, when the switch is forced by a move or ability. In this case, the switched mon is removed from game.scene.field by the leaveField method _before_ the SwitchSummonPhase, which is when these abilities are triggered. To accomodate this case, all abilities with a preSwitch trigger have been given a new bypassOnField attribute, which lets them work as intended (see for example the tests for Wimp Out and U-turn, in particular their interaction with Regenerator).
Note that a better fix might come from refactoring the order of calls in ForceSwitchOutHelper and dependent functions, to ensure that the leaveField method is always called _after_ all preSwitch abilities have triggered. This would remove the need for a bypassOnField attribute altogether, but would require significant intervention on the codebase.


## Screenshots/Videos

Before this PR (wonder guard doesn't come back on after Koffing is roared out):
https://github.com/user-attachments/assets/513388b7-d144-4550-bdd0-15f40d11fb85

After this PR (Neutralizing gas stops working, and Wonder Guard comes back up as expected):
https://github.com/user-attachments/assets/683d34fd-9ec0-4f0a-82b4-52b2275433e7


## How to test the changes?
Use the following override:
const overrides = {
  BATTLE_TYPE_OVERRIDE: "double",
  ABILITY_OVERRIDE: Abilities.WONDER_GUARD,
  MOVESET_OVERRIDE: [ Moves.ROAR, Moves.TACKLE, Moves.SPLASH ],
  OPP_SPECIES_OVERRIDE: Species.KOFFING,
  STARTING_BIOME_OVERRIDE: Biome.SLUM,
  OPP_MOVESET_OVERRIDE: [ Moves.SPLASH ],
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;

Go through battles until you find an encounters where one Koffing has Neutralizing Gas, while the other has a different ability. Use Roar on the Neutralizing Gas Koffing. After doing this, your active mons should be immune to Tackle as Wonder Guard is active again.


## Checklist
- [x ] **I'm using `beta` as my base branch**
- [x ] There is no overlap with another PR?
- [x ] The PR is self-contained and cannot be split into smaller PRs?
- [x ] Have I provided a clear explanation of the changes?
- [x ] Have I tested the changes manually?
- [x ] Are all unit tests still passing? (`npm run test`)
  - [x ] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?
- [x ] Have I provided screenshots/videos of the changes (if applicable)?

Are there any localization additions or changes? No
